### PR TITLE
feat(migrate): G5.3-T3 machine-local detector — path/hostname/username heuristic (#610)

### LINKED ISSUE #610: Task: G5.3-T3 Machine-local detector (path/hostname/username heuristic + per-pattern tests)
Parent initiative: #375 (G5.3 Laptop-local memory migration UX)
Parent goal: #216
Depends on: #608 (the `cli/internal/migrate/` package scaffold).
Independent of T2 (parser) — can be built and merged in parallel.

## Current state

After T1, `cli/internal/migrate/` exists but has no machine-local
heuristic. Nothing flags memory bodies that contain laptop-specific
paths, local-only hostnames, or the operator's username — so a naive
migration would leak `/Users/<name>/...` strings into `tenant`-scoped
memory where they mislead other operators (the failure mode #375's
Motivation section calls out).

## Desired state

`cli/internal/migrate/machinelocal.go` provides pure, cobra-free,
unit-testable detection:

- `DetectMachineLocal(body string) MachineLocalResult` where
  `MachineLocalResult` is `{ Flagged bool; Matches []MachineLocalMatch }`
  and `MachineLocalMatch` is `{ Category string; Sample string }`
  (category = which rule fired, sample = the matched substring,
  truncated to ~80 chars, for the T4 picker badge tooltip).
- Rules (each a distinct `Category`):
  - `unix-home`: `/Users/<seg>/` and `/home/<seg>/` (POSIX home
    prefixes; `<seg>` = one path segment).
  - `windows-home`: `C:\Users\<seg>\` (case-insensitive drive letter;
    backslash-escaped in the regex).
  - `tilde-home`: a `~/` token at a path boundary (start of string or
    preceded by whitespace/quote) to avoid matching `~/` inside prose
    like "approx ~/2 days".
  - `local-hostname`: hostnames ending `.local` or `.lan`, and the
    bare tokens `localhost`, `host.docker.internal`.
  - `operator-username`: the basename of `os.UserHomeDir()` appearing
    `>= 3` times in the body as a whole-word match (guard against a
    common-word username inflating false positives — document the
    3-occurrence threshold rationale in a comment; it mirrors the
    Initiative's spec).
- Whole-word / boundary matching uses Go `regexp` with explicit
  anchors — no third-party regex dep.
- `Flagged` is true iff `len(Matches) > 0`. The function is the
  single source of truth the T4 picker calls to pre-mark a file
  "Skip — machine-local"; `--include-machine-local` (T4) overrides
  the pre-mark, it does not change detection.

## Acceptance criteria

- [ ] `cd cli && go test ./internal/migrate/...` passes with a table-driven test that has at least one positive and one negative case per `Category`, including: `/Users/alice/x` (hit) vs `/Usersalice` (miss); `C:\Users\bob\` (hit); `~/foo` at line start (hit) vs `approx ~/3 weeks` mid-prose (miss); `db.local`/`box.lan`/`localhost`/`host.docker.internal` (hits) vs `mylocaldb` (miss); username at 2 occurrences (miss) vs 3 (hit).
- [ ] `operator-username` rule reads `os.UserHomeDir()` basename; test injects a fake home via a seam (function variable or parameter) so it is deterministic and does not depend on the CI runner's `$HOME`.
- [ ] Each `MachineLocalMatch.Sample` is non-empty for a hit and truncated to <= 80 chars.
- [ ] No cobra import in the file; `go vet` clean; `gofmt -d` empty.

## Out of scope

- Wiring detection into the picker UI / the "Skip — machine-local"
  default selection / `--include-machine-local` override — that is
  T4. T3 only provides `DetectMachineLocal`.
- The explicit `<!-- meho:machine-local -->` opt-out comment — that is
  T2's `scan.go` (`MachineLocalOptOut`); T3 is the *heuristic*, T2 is
  the *manifest opt-out*; both feed T4 independently.
- Inherits all Out-of-scope bullets of Initiative #375.

## References

- Parent: Initiative #375, Goal #216. The exact pattern list is
  specified in #375 work-item 2 and Definition-of-done bullet 5.
- Precedent for deterministic, seam-injected, table-driven internal
  tests: `cli/internal/auth/config_test.go`.
- Best practices: **ai_engineering_best_practices** §Safety
  (the heuristic exists to stop machine-local leakage into shared
  scopes — a privacy/correctness guardrail, default-deny: flagged
  files default to skip and require an explicit operator override).
- Go `regexp` — https://pkg.go.dev/regexp (stdlib; no new dep).

### PR DESCRIPTION
**Pull request by ddzafic**
## Summary

- Adds `DetectMachineLocal(body string, homeFn HomeDirFunc) MachineLocalResult` to `cli/internal/migrate/` — pure stdlib, no cobra import, no third-party regex dep.
- Five heuristic rules (unix-home, windows-home, tilde-home, local-hostname, operator-username) flag laptop-specific content so the T4 picker can pre-mark files "Skip — machine-local".
- `HomeDirFunc` seam makes the operator-username rule deterministic in CI — tests inject a fake home path rather than depending on the runner's `$HOME`.
- `MachineLocalMatch.Sample` is truncated to ≤80 runes for the T4 badge tooltip.

## Test plan

- [x] `cd cli && go test ./internal/migrate/...` — all 21 subtests pass
- [x] `go vet ./...` — clean
- [x] `gofmt -d` — empty diff
- [x] `go test ./...` — all 18 packages pass (0 failures)
- [x] `python3 .claude/hooks/code-quality.py --diff` — exit 0
- [x] unix-home: `/Users/alice/x` hit; `/Usersalice` miss
- [x] windows-home: `C:\Users\bob\` hit; no trailing backslash miss
- [x] tilde-home: `~/foo` at line start hit; `approx ~/3 weeks` mid-prose miss
- [x] local-hostname: `db.local` / `box.lan` / `localhost` / `host.docker.internal` hits; `mylocaldb` miss
- [x] operator-username: 3 occurrences hit; 2 occurrences miss; error from homeFn miss; nil homeFn no panic
- [x] Each `MachineLocalMatch.Sample` non-empty and ≤80 runes

## Out of scope

- T4 picker wiring (`--include-machine-local` flag, per-file UI) — that is #612.
- Frontmatter parser / scanner (`scan.go`) — that is #609 (parallel sibling).
- `<!-- meho:machine-local -->` HTML-comment opt-out — that is T2 (#609).

Closes #610

## Release Notes

* **New Features**
  * Added machine-local pattern detection to identify home directory paths, usernames, and hostname indicators in document text.

* **Tests**
  * Added comprehensive test coverage for detection patterns and edge cases.

* **Documentation**
  * Updated module documentation for the new detection capability.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/644?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by ddzafic** [comment]
@copilot review

**Comment by ddzafic** [comment]
@copilot review

File: cli/internal/migrate/machinelocal.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	home, err := homeFn()
+	if err != nil || home == "" {
+		return nil
+	}
+	username := filepath.Base(home)
+	if username == "" || username == "." || username == string(filepath.Separator) {
+		return nil
+	}
+
+	pattern := fmt.Sprintf(`(?i)\b%s\b`, regexp.QuoteMeta(username))
```
Comment (damir-topic) [inline cli/internal/migrate/machinelocal.go]: **[⚠️ Major]** — `operator-username` rule silently never fires for non-ASCII OS usernames

**What:** `pattern := fmt.Sprintf(`(?i)\b%s\b`, regexp.QuoteMeta(username))` uses Go's `\b`. In Go's `regexp` (RE2) `\b` and `\w` are **ASCII-only** — `\w == [0-9A-Za-z_]`. For a username like `josé`, `müller`, `владимир`, the trailing `\b` is evaluated between `é` (non-`\w` under the ASCII definition) and a following space (also non-`\w`): no `\w`/`\W` transition, so `\b` does not hold and the token fails to match. `FindAllString` returns 0 hits → `len(hits) < 3` → `detectUsername` returns `nil`.

**Why it matters:** The other four rules are username-independent and still work, but this rule is the one that catches `/Users/<name>/…`-class leakage where the path segment *is* the operator's name. For any operator whose OS account name contains a non-ASCII letter (common across the European team this ships to), the rule is a silent no-op — exactly the "leak `/Users/<name>/...` into tenant-scoped memory" failure #375's Motivation calls out, and a default-deny safety guardrail (`ai_engineering_best_practices §Safety`) that fails open. The acceptance test only exercises `alice`, so CI stays green while the contract is broken for a whole class of users. CodeRabbit independently flagged the same line; I concur and rate it Major (not a nitpick) on the safety rationale.

**Required change:** Use a Unicode-aware boundary and capture the token so `Sample` excludes the boundary char:

```go
pattern := fmt.Sprintf(`(?i)(?:^|[^\pL\pN_])(%s)(?:$|[^\pL\pN_])`, regexp.QuoteMeta(username))
re, err := regexp.Compile(pattern)
if err != nil {
	return nil
}
hits := re.FindAllStringSubmatch(body, -1)
if len(hits) < 3 {
	return nil
}
matches := make([]MachineLocalMatch, 0, len(hits))
for _, h := range hits {
	if len(h) < 2 || h[1] == "" {
		continue
	}
	matches = append(matches, MachineLocalMatch{
		Category: "operator-username",
		Sample:   truncate(h[1], 80),
	})
}
```

Add a regression case to the `operator-username` table: a non-ASCII username (e.g. `fakeHome("/Users/josé")`, body with `josé` ×3) must `match: true`. (Note: with a consuming boundary class, adjacent occurrences sharing one separator can under-count; the `josé ran josé then josé` shape used in the existing test has distinct separators and is unaffected — but verify the 3-occurrence case still trips after the change.)

**Source:** https://pkg.go.dev/regexp/syntax — “`\b` at ASCII word boundary … `\w` word characters (== `[0-9A-Za-z_]`)”; confirmed via fetch against the live doc.

**Review by damir-topic — CHANGES_REQUESTED** [review]
## /review-pr-slim summary

**Anchor task:** #610 — G5.3-T3 machine-local detector (path/hostname/username heuristic)
**Surface area:** backend (Go CLI — pure stdlib, no cobra)
**Best-practices packs audited:** ai_engineering (§Safety — default-deny privacy guardrail); no Go-specific pack in repo, reviewed against Go `regexp`/RE2 semantics (cited)
**Verdict:** REQUEST_CHANGES

### Acceptance criteria
- [x] `go test ./internal/migrate/...` table-driven, ≥1 ± per Category incl. all spec examples (`/Users/alice/x` vs `/Usersalice`; `C:\Users\bob\`; `~/foo` line-start vs `approx ~/3 weeks`; `db.local`/`box.lan`/`localhost`/`host.docker.internal` vs `mylocaldb`; username 2 vs 3) — verified, 21 subtests pass on head `7cd7f44`
- [x] `operator-username` reads `os.UserHomeDir()` basename via injected `HomeDirFunc` seam — deterministic, CI-independent
- [x] each `Sample` non-empty for a hit and ≤80 — `TestSampleTruncation` (rune-based) verifies
- [x] no cobra import; `go vet` clean; `gofmt -d` empty — verified in worktree
- ⚠️ **Criteria under-specify the rule's contract:** no non-ASCII-username case is required, so the `\b` safety hole below passes CI while the rule fails open for non-ASCII operators (see Major).

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #610 | open (not on a project board) | `Closes #610`, base `main` | would fire on merge to `main` — **but PR is unmergeable** | ⚠️ code complete, 1 Major + structural blocker | resolve blocker (rebase/retarget) + Major, then close on merge |

T3 is "independent of T2 — mergeable in parallel" per the task; no orphan/collision with sibling PRs #645 (T2/#609), #680 (T4/#611), #685 (T5).

### Predicted vs. actual diff
- ✅ `machinelocal.go` — `DetectMachineLocal` + 5 rules match #610; RE2 regexes are linear-time (no ReDoS — Go `regexp` is NFA/RE2)
- ✅ `machinelocal_test.go` — per-Category ±, truncation, seam (`fakeHome`/`fakeHomeErr`/nil), flagged-consistency
- ✅ `docs/codebase/cli.md` (T3 delta) — accurate, no .NET analogies
- ✅ detector correctly **not** wired into `cmd/` (T4 scope — verified by grep)
- 🚫 **8 unmerged stack commits bundled** (#608 scaffold, #640 CI ×5, #679 conftest) — see structural Blocker; not the author's T3 work, an artifact of basing on `main` instead of the #608 branch

### Integration verification (all in worktree on head `7cd7f44`)
- [x] `go vet ./internal/migrate/...` — clean
- [x] `gofmt -d ./internal/migrate/` — empty
- [x] `go test ./internal/migrate/...` — 21 subtests pass; `go test ./...` — all 20 packages pass, no regressions
- [x] `git merge-tree origin/main HEAD` — **CONFLICT in `docs/codebase/cli.md`** (drives `mergeable=false/dirty`)
- [x] detector-wiring grep — `DetectMachineLocal` referenced only in its own pkg+test (T4 scope correctly excluded)

### Findings (inline — each is a merge blocker per team policy)
- 🛑 1 blocker — **structural, summary-level (no code line to anchor):** PR `mergeable=false (dirty)`. Conflict in `docs/codebase/cli.md` vs `main`; base is `main` while #610 `Depends on: #608`, so the PR carries 8 unmerged stack commits and its diff view is not reviewable in isolation. **Required:** retarget the PR base to the #608 stack branch (`feat/608-migrate-cmd-huh-v2`, as sibling #645/T2 does) and rebase; or rebase onto `main` after #608/#640 land and resolve the `cli.md` conflict. **No T3 code change required for this item.**
- ⚠️ 1 major — `\b` ASCII-only → `operator-username` silently no-ops for non-ASCII usernames (inline on `machinelocal.go`; concurs with CodeRabbit's machinelocal.go:123 finding, elevated to Major on the safety rationale)
- 💡 0 minors
- ❓ 0 questions

### Bot review
CodeRabbit ran (base=`main` is the default branch, so its auto-review is enabled here — unlike sibling #645). Its one actionable inline (machinelocal.go:123, `\b` Unicode) is **valid** — independently confirmed against pkg.go.dev/regexp/syntax. Its test-addition suggestion was filed as a "🧹 nitpick / quick win"; the underlying defect is Major, not optional — addressed in the inline above so it is not dismissed as a nit by the auto-fix loop. No Copilot review (not requested).

### Adjacent improvements (non-blocking, not posted inline)
- `tilde-home` `Sample` includes the leading boundary char it matched (`\n~/`, `"~/`) because `FindAllString` returns the whole `(?:^|[…])~/` match; cosmetic for the T4 tooltip, harmless. A capture group (as the username fix uses) would trim it — worth a future consistency pass, not required here.
- `detectUsername` emits one `MachineLocalMatch` per occurrence (N matches for N hits); consistent with the other rules' `FindAllString` behavior and fine for T4's needs — dedup-to-one is a possible future tidy, not required.

### Docs
- [x] `docs/codebase/cli.md` — T3 delta accurate and present; **but** conflicts with `main` (the structural blocker). No .NET analogies.

### What worked well
- Faithful, minimal, spec-accurate: all 5 rules map 1:1 to #610, RE2-only (no third-party regex dep), and the `HomeDirFunc` seam makes the username rule deterministic exactly as the criteria asked.
- Test table is honest — it asserts the *misses* the spec calls out (`/Usersalice`, `approx ~/3 weeks`, `mylocaldb`, 2-occurrence username), not just the hits.
